invenio_app_ils/anonymization.py

```diff
@@ -0,0 +1,193 @@
+from invenio_app_ils.document_requests.api import DocumentRequest
+from invenio_app_ils.errors import PatronNotFoundError
+from invenio_app_ils.ill.api import BorrowingRequest
+from invenio_app_ils.ill.proxies import current_ils_ill
+from invenio_app_ils.patrons.api import (Patron, SystemAgent,
+                                         get_anonymous_patron_dict,
+                                         get_patron_or_unknown)
+
+from .acquisition.search import OrderSearch
+from .circulation.search import get_loans_by_patron_pid
+from .document_requests.search import DocumentRequestSearch
+from .ill.search import BorrowingRequestsSearch
+from .patrons.indexer import PatronIndexer
+from .proxies import current_app_ils
+
+
+def get_patron_activity(patron_pid):
+    """Get activity related to the given patron pid."""
+    if patron_pid is None:
+        raise ValueError("No patron pid was provided.")
+
+    patron = get_patron_or_unknown(patron_pid)
+    if not patron:
+        return None
+
+    def dump(search):
+        return [hit.to_dict() for hit in search.scan()]
+
+    patron_document_requests = dump(
+        DocumentRequestSearch().search_by_patron_pid(patron_pid)
+    )
+
+    patron_borrowing_requests = dump(
+        BorrowingRequestsSearch().search_by_patron_pid(patron_pid)
+    )
+
+    patron_acquisitions = dump(OrderSearch().search_by_patron_pid(patron_pid))
+
+    patron_loans = dump(get_loans_by_patron_pid(patron_pid))
+
+    patron_profile = UserProfile.get_by_userid(patron_pid).__dict__
+    del patron_profile["_sa_instance_state"]
+
+    patron_data = {
+        "patron": patron,
+        "profile": patron_profile,
+        "document_requests": patron_document_requests,
+        "borrowing_requests": patron_borrowing_requests,
+        "acquisitions": patron_acquisitions,
+        "loans": patron_loans,
+    }
+
+    return patron_data
+
+
+def anonymize_patron_data(patron_pid, force=False):
+    """Anonymize all the data/activity related to the given patron pid."""
+    if patron_pid is None:
+        raise ValueError("No patron pid was provided.")
+
+    patron = get_patron_or_unknown(patron_pid)
+    if not patron and not force:
+        return None
+
+    patron_object = None
+    try:
+        patron_object = current_app_ils.patron_cls.get_patron(patron_pid)
+    except PatronNotFoundError:
+        if not force:
+            raise PatronNotFoundError(patron_pid)
+
+    if (
+        get_loans_by_patron_pid(patron_pid)
+        .filter("term", state="ITEM_ON_LOAN")
+        .count()
+    ):
+        raise AssertionError(
+            "Cannot delete user %s: they have ongoing loans." % (patron_pid)
+        )
+
+    # Serialize empty patron values
+    anonymous_patron_fields = get_anonymous_patron_dict(patron_pid)
+
+    patron_loans = get_loans_by_patron_pid(patron_pid).execute()
+
+    indices = 0
+
+    for hit in patron_loans:
+        loan = Loan.get_record_by_pid(hit.pid)
+        if loan["state"] == "PENDING":
```

Comment:
it would be better to use (coming from invenio-circulation)
```
current_app.config["CIRCULATION_STATES_LOAN_REQUEST"] 
```
instead of hardcoding the state

Reply:
You're right, thanks; I would also like to point out that most of the existing code "hardcodes" the states (thus refactoring might be needed).